### ISSUE TITLE
[Request] JiffyDOS, is it possible ?

### ISSUE THREAD

**Issue by WindyWinston** [libretro/vice-libretro#215]
Hi, is it currently possible to get JiffyDOS working on this Vice ? In WinVice you need to set the path for the Kernal ROM e.g. C64 and the path to the Disk Drive ROM e.g. 1541-II and that's it. Can this currently be done and if not, would it be easy to implement ? 

JiffyDOS seems to be very popular with C64 users as it has much faster loading and much higher compatibility than using a cartridge method like Action Replay.

I saw this issue https://github.com/libretro/vice-libretro/issues/164 which JiffyDOS can do. I think it was F1 to list disk contents, move to the required file, then F5 and enter to run it (or the UP arrow).

**Comment by sonninnos** [comment]
Yep, seems to be working:
![retroarch_2020_03_25_15_02_11_834](https://user-images.githubusercontent.com/45124675/77539300-d0cf5b00-6ea9-11ea-80a8-b8130356413a.png)

`RA/system/vice/vicerc` acts as vice.ini, or whatever it is named, so simply copy the rows from there. The result should be like:

```
[C64SC]
KernalName="x"
DosName1541ii="y"
DosName1581="z"
etc..
```

Perhaps this current way is enough for now, but why not a core option too if it is worth it.

And thanks for the reminder, I'll be surely testing more with this. Especially as it also works with 1581, unlike that Action Replay method.

**Comment by WindyWinston** [comment]
This is awesome ! Seems a much better alternative to the Action Replay and also means the Cartridge slot is free for anything else.

Apart from 1581, what drive type does this Vice use ? `1541` or `1541-II` ? I'm wondering what ROM to use. As you probably know a lot about this, could you give the md5 of the best/most compatible ROMS ?

**Comment by sonninnos** [comment]
Seems to default to 1541. Those rows mentioned above are enough for the obvious difference.

Too early to say anything about compatibility, but I found these to be working (same file):
KernalName = `JiffyDOS_Kernal.bin` - md5: `be09394f0576cf81fa8bacf634daf9a2`
KernalName = `JiffyDOS_C64.bin` - md5: `be09394f0576cf81fa8bacf634daf9a2`

Both of these work (again same file), but `JiffyDOS_1541.bin` does not:
DosName1541 = `JiffyDOS_C1541.bin` - md5: `1b1e985ea5325a1f46eb7fd9681707bf`
DosName1541 = `JiffyDOS_1541-II.bin` - md5: `1b1e985ea5325a1f46eb7fd9681707bf`

DosName1581 = `JiffyDOS_1581.bin` - md5: `20b6885c6dc2d42c38754a365b043d71`

**Comment by WindyWinston** [comment]
I'll stick with the vanilla 1541 for compatibility.

**Comment by WindyWinston** [comment]
My C64 JiffyDOS Kernal and 1581 have the same md5 as you, but for 1541 mine is different. I ran them through TOSEC and all the files they have are labelled [outdated]. 

The md5 you listed is named as -
`1541-II JiffyDOS VX.XX (19xx)(Creative Micro Designs)[outdated]`

Mine says -
`1541 JiffyDOS VX.XX (19xx)(Creative Micro Designs)[outdated]`
md5: `0613772F604563A25A619940938F9C67`

I wonder if there is a recent Drive ROM out there. Also, is there a way to know if the Drive ROM is actually working correctly ?

**Comment by sonninnos** [comment]
That is the same `JiffyDOS_1541.bin` that would not work.

**Comment by WindyWinston** [comment]
So the one you listed which is for `1541-II` is the way to go? But is this correct because you said that this Vice is using the vanilla 1541 drive ?

**Comment by sonninnos** [comment]
This is the only way I can get it to work, so that is correct enough for me.

The current disk image format autodetection sets the drive to 1541, therefore understandably only `DosName1541=` works. If it would be set to 1541-II then it would be named `DosName1541ii=`. 

As for the ROM being named for 1541-II and it still works with 1541: Jiffy authors might know what is up. I don't see a reason to do anything about it, as this is fine.

**Comment by WindyWinston** [comment]
Yeah, this is good. I just had a check with 'Mayhem in Monsterland', Warp off, and roughly timed from the load message until the screen turned black. With the 1541 rom it took about 26 seconds. The 1541-II rom took 8 seconds! A huge difference.

**Comment by WindyWinston** [comment]
On my googleing I found a post in a Retropie forum about JiffyDOS and they added a couple more lines to the 'vicerc' file, not sure what they do -
```
KernalRev=-1
Drive8IdleMethod=1
```

There's also talk of another kernal called DolphinDOS 3 which is supposed to be even faster. I don't think it has 1581 support and not sure of the compatibility but it has a memory expansion. I've yet to dabble. Here's the link to the post -
https://retropie.org.uk/forum/topic/25152/lr-vice-how-to-use-alternate-kernal-like-jiffydos

**Comment by WindyWinston** [comment]
So I thought I'd try a demo on this JiffyDOS thing. The first demo I tried, 'Krestage'. Didn't get past the Title screen, black screen! I don't think the compatibility is all its cracked up to be. That is only the second thing I've tried so far. I must have a knack for finding these things.

**Comment by sonninnos** [comment]
KernalRev seems to be -1 already. IdleMethod requires more investigating.

**Comment by WindyWinston** [comment]
I checked out a couple of others, S-JiffyDOS and DolphinDOS. They are both faster but apparently less compatible so I think JiffyDOS is the only one to use (and most popular). If you type @ (then enter) at the DOS screen it shows the version of the Drive ROM. It says v5.0 which I think is the latest as i've seen the same on a few Youtube vids.

Do you wan't me to leave this thread open for any reminders ?

**Comment by sonninnos** [comment]
`JiffyDOS_1541-II.bin` appears to be 5.0 and `JiffyDOS_1581.bin` says 6.0.

I'll add a core option soonish, so sure leave it open. I'll go with the above mentioned files, but there is still time to chime in and suggest better ones.

My collection only has 2 games that need Jiffy disabled. The number is 10 with Action Replay. Didn't fool around with that IdleMethod yet.

**Comment by WindyWinston** [comment]
Yeah, I think your original post of vicerc config & md5's was perfect -

```
KernalName = JiffyDOS_C64.bin - md5: be09394f0576cf81fa8bacf634daf9a2
DosName1541 = JiffyDOS_1541-II.bin - md5: 1b1e985ea5325a1f46eb7fd9681707bf
DosName1581 = JiffyDOS_1581.bin - md5: 20b6885c6dc2d42c38754a365b043d71
```
What would be the core option as we already have the one to use the vicerc file ?

And what were the 2 games you found that dont work ? I haven't had time to test anything yet but when I tried 1 game and 1 demo I found the 'Krestage' demo didn't work.

**Comment by sonninnos** [comment]
The core option of course would simply remove the manual labor. Check for the files under `system/vice` & override vicerc. Currently the vicerc core option isn't resetting all necessary settings properly on any kind of restart anyway, so that needs some attention too.

The games are "Bobix" & "Crazy Cars III". Maybe there are other versions that do work.

**Comment by WindyWinston** [comment]
Some CRT's don't work with JiffyDOS either ! For example -
[gauntlet-dd.zip](https://github.com/libretro/vice-libretro/files/4397556/gauntlet-dd.zip)

There's always something that ruins the experience !

**Comment by sonninnos** [comment]
Yeah, using Jiffy for disk images only sounds like a good idea. What there is to gain and why when using PRG/CRT.. As in looks like the old TODO got an addition then.

**Comment by sonninnos** [comment]
Derp, should have put some more thought to it, as of course this current whitelist method is not good for M3Us full of D64s and starting the core without content.. I'll change it soonish to just blacklist PRG & CRT instead.

**Comment by WindyWinston** [comment]
Good thinking, Batman. As TAP's and T64's don't work also because they need a Tape device, would it be good to blacklist those as well ?

**Comment by sonninnos** [comment]
Indeed TAPs will not load at all with Jiffy. Took a moment to figure out how to allow D64s in M3Us but not TAPs in M3Us, but seems to be sorted now.

**Comment by sonninnos** [comment]
Score! Found Jiffy-compatible versions of those above mentioned games from the good old place.

**Comment by WindyWinston** [comment]
That 1541 ROM that didn't work is only half (md5: `0613772F604563A25A619940938F9C67`). Apparently you replace the second half of the dos1541 ROM (from $2000 onwards) with that Jiffy 1541 ROM (8k). I tried this and the resulting md5 is the same as the `JiffyDOS_1541-II.bin` that we are using now (md5: `1b1e985ea5325a1f46eb7fd9681707bf`).

Just thought i'd let you know.

**Comment by WindyWinston** [comment]
As C128 is getting some action would it be good to enable JiffyDOS for it ?

Are these two the correct ones ?
CBBD1BBCB5E4FD8046B6030AB71FC021  JiffyDOS_C128.bin
41C6CC528E9515FFD0ED9B180F8467C0  JiffyDOS_1571.bin

Not sure if the 'GO 64' command would work with it though ?

**Comment by sonninnos** [comment]
All help is welcome. You can scout out how that works with `vicerc`. 1571 could be also added, sure why not, even though there aren't many disks around needing it.

**Comment by WindyWinston** [comment]
What drive does the C128 use currently in this core ? I thought the 1571 was the Commodore 128 disk drive ?

**Comment by sonninnos** [comment]
Oh right, C128 defaults to 1571 in standalone, even though it matters little here since the used image type is what ultimately decides the drive.

I'll add the support for C64 too while I'm at it.

**Comment by sonninnos** [comment]
Don't know what is up with that, I also wasn't able to get the floppy version of the newish SMB port to load, but the cartridge vworked fine. Some other random disks, autoboot or not started fine. Prince of Persia also runs nice.

I decided not to rename the 1571 ROM to a cleaner name, even if it does look ugly (`JiffyDOS_1571_repl310654.bin`). It has the same MD5 though. `JiffyDOS_1571D.bin` did not work.

**Comment by WindyWinston** [comment]
On C128, in JiffyDOS, if I type `@` to show the Jiffy Drive version it says 'Drive Not Ready' ?

**Comment by sonninnos** [comment]
Argh, for some reason on C128 the current method of disabling JiffyDOS when running tapes inside a M3U does not work like it does in the C64 core.

What disk are you trying? I did find some oddities in the disks I found, like D64's that have something that makes them autoboot without the usual macro (C128 does Amiga-style boot check), but DirMaster shows no extra files. The autobooting disk was a conversion of Qwak.

The SMB I tested is in D64+CRT and I don't have any T64 at hand right now, just some TAPs, but they should behave just the same in this context. Yes, the disk image format changes the drive type accordingly, and no other drive type sets are involved except the 1571 being the default when starting the core without content.

![retroarch_2020_04_28_06_45_45_372](https://user-images.githubusercontent.com/45124675/80445322-a6474680-891c-11ea-9888-f2707501bca8.png)
![retroarch_2020_04_28_06_46_08_539](https://user-images.githubusercontent.com/45124675/80445325-a8110a00-891c-11ea-9d08-04e7682e3ba5.png)
![retroarch_2020_04_28_06_46_56_488](https://user-images.githubusercontent.com/45124675/80445330-aa736400-891c-11ea-8345-d79f35a3e94e.png)

**Comment by WindyWinston** [comment]
I downloaded a few 128 only games from CSDB. All of them kept autobooting after a reset except Kikstart 2 128 (d64) but when I tried the `@` it said something like `00, OK, 00`.

The Super Mario Bros 64 I have is a .t64 trained +5 by Excess. After loading it says BREAK and lots of number codes.

To be honest i'm just going to stick with the superior C64. There is only a handful of C128 only games anyway.
